rfc-117-containerised-runtime.md

```diff
@@ -0,0 +1,68 @@
+- Our current version of Puppet is 3.8, which is no longer supported (the current version is 6.13). Support for newer operating system versions is likely to be lacking in unsupported versions of Puppet.
+- Other infrastructure as code tools in use (e.g. [Fabric](https://github.com/alphagov/fabric-scripts)) are also currently using very old versions which are very likely to have problems with newer operating systems.
+
+## Proposal
+
+GOV.UK will adopt a container based hosting environment, such as [GOV.UK PaaS](https://www.cloud.service.gov.uk). The choice of hosting environment is out of scope for this RFC.
+
+GOV.UK applications will be built into container images, which will include the operating system. This may be done using buildpacks or other means - this decision is also out of scope for this RFC.
+
+A containerised runtime aligns with the [GOV.UK Infrastructure Architecture Goals for 2021](
+https://docs.google.com/document/d/1ooN7wkYhEGvceGe9Qz_HNZa-GPtrjzK_vA4vfWYVn4c/edit#heading=h.cdrr7rv9t98f) to isolate applications through containers. This will give us better control of how resources are matched to applications.
+
+The TechOps strategy is to use common components where available. GOV.UK has been following the strategy by using hosted versions of software ([ElasticSearch](https://aws.amazon.com/elasticsearch-service/) and [Postgres](https://aws.amazon.com/rds/)) and services ([Notify](https://www.notifications.service.gov.uk)) where available. Using a containerised runtime is a continuation of this policy.
+
+GOV.UK has already containerised its development environment ([rfc-106](https://www.github.com/alphagov/govuk-rfcs/106)), a containerised runtime will narrow the gap between development and production.
+
+### Benefits of this approach
+
+The most immediate benefit of this approach is that it will allow us to incrementally upgrade the operating system version. By moving the applications into containers, GOV.UK can enable a much lower risk upgrade path.
+
+Moving to a container based hosting environment will allow us to remove much of our legacy Puppet and Fabric code. This code is hard to maintain, and hard to hire people with the experience required to improve and upgrade. With industry and government overwhelmingly moving towards containerised platforms, it should be much easier to hire people with these skills.
+
+Containerised hosting environments can generally provide much better value for money in terms of resource usage, both because many small applications can be “bin-packed” onto the same infrastructure, and because it is much easier to autoscale applications.
+
+Running containers in production allows for a reduction in the difference between development environments and production. This can lead to bugs being found more quickly, and faster development cycles.
```

Comment:
Does this cater for everyone working on GOV.UK? In theory you'll find bugs faster and more consistently because environments are similar, but different people use different tools when developing and diagnosing bugs. For example, how easily can an IDE be used to debug a Docker application? We might need to do some more investigation and put more effort into this first.

Reply:
Isn't the emphasis that production bugs are found more quickly as they are more likely to show up and be reproduced in your development environment? Regardless of how you go about debugging them.

Reply:
I'd suggest that this benefit may be rather dependant on the decisions about the type of platform this is hosted on. For example if we use a build pack based approach for production there's likely to be no reduction in differences between dev and prod.

Reply:
Yeah, this point hides a few assumptions about the type of platform, how containers are built, and where they fit in to the development experience, all of which are out of scope for the RFC.

That said, the statement that "Running container in production _allows_ for a reduction in the difference" is still true, even if we decide not to do much with production-like development environments.

@theseanything mentions testing configuration locally, which I think is a good example of the benefit of being able to play with things in a production-like configuration outside of a full environment. Perhaps we could add that to this point:

```suggestion
Running containers in production allows for a reduction in the difference between development environments and production. For example, it can allow developers to test some configuration changes locally, without having to deploy to an environment. This can lead to bugs being found more quickly, and faster development cycles.
```

Reply:
I've added the above suggestion in c8170ef6c0d727b6638d83849a1263c2d5769cfd.

I'm not sure that fully resolves @karlbaker02 and others comments in this thread though.

I'd say that this RFC isn't proposing a change to the experience of writing code or debugging local apps - people would still do that with govuk-docker or however they personally prefer. That process won't get any better with containers in production.

What _might_ get better is the ability to do `docker-compose up` (or similar) with the versions of the containers that are run in production. As @kevindew says, whether that's really practicle rather depends on some choices we haven't made yet (e.g. currently buildpacks are less suitable for this use case).

I think the current wording ("allows for a reduction") is vague enough to suggest that this is a benefit we may or may not see depending on other choices.

Does that resolve this question?